<a href="https://colab.research.google.com/github/LC1332/personality-text-generation/blob/main/lulu_exp/%E5%B0%86%E5%88%86%E6%95%B0%E6%89%93%E5%8D%B0%E4%B8%BA%E8%A1%A8%E6%A0%BC%E6%88%96%E8%80%85%E5%8F%AF%E8%A7%86%E5%8C%96.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
fname_response = '/content/drive/MyDrive/Psy/psy_score_openai.jsonl'

In [3]:
import json

datas = []

with open(fname_response, encoding='utf-8') as f:
    for line in f:
        data = json.loads(line)
        datas.append(data)

In [4]:
print(len(datas))
print(datas[0])

32
{'role_name': '汤师爷', 'N': {'score': 0.4, 'sub_factors': {'N4': 0.4, 'N1': -5.0, 'N3': 0.4, 'N2': 5.0, 'N6': 1.75}}, 'C': {'score': 0.0, 'sub_factors': {'C4': 1.0, 'C2': -0.2857142857142857, 'C1': 1.0, 'C5': -1.0, 'C3': -0.6666666666666666, 'C6': 0.0}}, 'A': {'score': -0.25, 'sub_factors': {'A2': -0.3333333333333333, 'A4': -0.3333333333333333, 'A1': -0.3333333333333333, 'A3': 0.0, 'A5': 0.0}}, 'O': {'score': -0.2857142857142857, 'sub_factors': {'O5': -0.4, 'O3': 0.0, 'O1': 2.0, 'O2': -0.4, 'O4': 0.0, 'O6': -1.0}}, 'E': {'score': -2.0, 'sub_factors': {'E6': -2.0, 'E2': -2.0, 'E4': -2.0, 'E1': -2.0, 'E3': -2.0}}, 'qids': [[20, 75, 77, 69], [25, 40, 50, 70], [30, 45, 55], [35, 72, 73], [71, 76, 74], [1, 78, 36, 81], [41, 21, 80, 85], [82, 79, 84], [6, 83, 46], [11, 26, 51], [17, 31, 56], [2, 62, 47], [7, 27, 65], [12, 32, 66], [16, 63, 52], [60, 64, 57], [61, 37, 67], [22, 42, 68], [3, 43, 38], [8, 88, 48], [33, 90, 58], [86, 94, 93], [87, 18, 95], [92, 28, 97], [13, 89, 53], [23, 91, 9

In [5]:
long_name = {
    'N':'neuroticism',
    'E':'extraversion',
    'O':'openness',
    'A':'agreeableness',
    'C':'conscientiousness'
}

In [ ]:
print('角色', end = ' | ')

for key in long_name.keys():
    print(long_name[key], end = ' | ')

print()

print('|-|',end = '')

for key in long_name.keys():
    print('-', end = ' | ')

print()


for data in datas:
    role_name = data['role_name']

    print('|', role_name, end = ' | ')

    for key in long_name.keys():
        major_score = data[key]['score']

        print('%.3f' % major_score, end = ' | ')

    print()



角色 | neuroticism | extraversion | openness | agreeableness | conscientiousness | 
|-|- | - | - | - | - | 
| 汤师爷 | 0.400 | -2.000 | -0.286 | -0.250 | 0.000 | 
| 慕容复 | -3.800 | 1.167 | 1.286 | -0.750 | 3.375 | 
| 李云龙 | -2.000 | -0.667 | -2.857 | -1.000 | 1.500 | 
| Luna | -0.400 | 2.000 | 5.000 | 2.250 | 0.125 | 
| 王多鱼 | -2.600 | 3.500 | 0.714 | -1.500 | -1.000 | 
| Ron | -0.600 | -0.833 | -1.429 | 1.125 | -0.750 | 
| 鸠摩智 | -2.000 | -2.000 | 1.333 | -0.250 | 2.500 | 
| Snape | -1.200 | -2.500 | -2.000 | -2.000 | 1.750 | 
| 凉宫春日 | -2.600 | 2.833 | 3.714 | -1.500 | 1.500 | 
| Malfoy | -2.000 | -2.000 | -2.000 | -2.000 | 1.750 | 
| 虚竹 | 1.200 | -2.500 | 0.714 | 4.250 | nan | 
| 萧峰 | -2.000 | -0.167 | -0.857 | 1.500 | 3.375 | 
| 段誉 | -1.800 | 2.167 | 1.714 | 4.625 | 0.875 | 
| Hermione | -2.000 | 0.000 | 4.571 | 2.000 | 4.250 | 
| Dumbledore | -2.000 | nan | 4.571 | 3.125 | 2.625 | 
| 王语嫣 | 1.000 | -0.333 | 2.857 | 3.000 | 2.750 | 
| Harry | -0.400 | nan | -0.286 | 2.500 | 3.750 | 
| McGonag

long_name = {
    'N':'neuroticism',
    'E':'extraversion',
    'O':'openness',
    'A':'agreeableness',
    'C':'conscientiousness'
}

datas是一个list of json

其中每个元素一定包含role_name字段

检查datas的每一个元素，是否包含N_table, E_table等对应的table字段

如果没有，则打印 <role_name>不包含<X_table>

用python为我实现

In [6]:
long_name = {
    'N':'neuroticism',
    'E':'extraversion',
    'O':'openness',
    'A':'agreeableness',
    'C':'conscientiousness'
}

for data in datas:
    role_name = data['role_name']
    for short_name, long_factor in long_name.items():
        table_name = f'{short_name}_table'
        if table_name not in data:
            print(f'{role_name}不包含{table_name}')

{"role_name": "Dumbledore", "N": {"score": -2.0, "sub_factors": {"N4": -2.0, "N1": -2.0, "N3": -2.0, "N2": -2.0, "N6": -2.0}}, "C": {"score": 2.625, "sub_factors": {"C4": 3.5, "C2": 2.7142857142857144, "C1": 3.5, "C5": 1.75, "C3": 0.6666666666666666, "C6": 3.5}}, "A": {"score": 3.125, "sub_factors": {"A2": 3.5, "A4": 3.5, "A1": 3.5, "A3": 2.0, "A5": 2.0}}, "O": {"score": 4.571428571428571, "sub_factors": {"O5": 4.4, "O3": 3.5, "O1": 5.0, "O2": 5.0, "O4": 5.0, "O6": 4.25}}, "E": {"score": NaN, "sub_factors": {"E4": 2.6, "E1": 2.0}}, "qids": [[25, 40, 50, 70], [30, 45, 55], [35, 72, 73], [71, 76, 74], [1, 78, 36, 81], [41, 21, 80, 85], [82, 79, 84], [6, 83, 46], [11, 26, 51], [17, 31, 56], [2, 62, 47], [7, 27, 65], [12, 32, 66], [16, 63, 52], [60, 64, 57], [61, 37, 67], [22, 42, 68], [3, 43, 38], [8, 88, 48], [33, 90, 58], [86, 94, 93], [87, 18, 95], [92, 28, 97], [13, 89, 53], [23, 91, 96], [4, 99, 106], [24, 100, 107], [49, 105, 54], [98, 29, 101], [9, 34, 104], [14, 103, 108], [19, 44, 59], [39, 102, 109]], "N_table": {"N4": [-2, -2, -2, -2, -2], "N1": [-2, "null", "null", "null", "null"], "N3": [-2, -2, -2, -2, -2], "N2": ["null", -2, "null", "null", "null"], "N6": ["null", -2, -2, -2, -2]}, "C_table": {"C4": [2, 5, "null", "null", "null", "null", 2, 5], "C2": [2, 5, 5, "null", -2, 2, 2, 5], "C1": [2, 5, 5, 2, "null", "null", "null", "null"], "C5": ["null", "null", 5, 2, -2, 2, "null", "null"], "C3": ["null", "null", "null", 2, -2, 2, "null", "null"], "C6": ["null", "null", "null", "null", "null", "null", 2, 5]}, "A_table": {"A2": [5, 5, 2, 5, "null", "null", 2, 2], "A4": [5, 5, 2, 5, 2, 2, "null", "null"], "A1": [5, 5, 2, 5, 2, 2, "null", "null"], "A3": ["null", "null", "null", "null", 2, 2, 2, 2], "A5": ["null", "null", "null", "null", "null", "null", 2, 2]}, "O_table": {"O5": [5, 5, 5, "null", "null", 5, 2], "O3": [5, "null", "null", "null", "null", "null", 2], "O1": [5, "null", "null", "null", "null", "null", "null"], "O2": ["null", 5, 5, 5, 5, 5, "null"], "O4": ["null", 5, 5, 5, 5, "null", "null"], "O6": ["null", "null", "null", 5, 5, 5, 2]}, "E_table": {"E6": [2, 2, "null", "2", 2, 5], "E2": [2, "null", "null", "2", 2, 5], "E4": [2, 2, 2, "null", 2, 5], "E1": [2, 2, 2, "null", "null", "null"], "E3": ["null", 2, 2, "2", "null", "null"]}}


In [8]:
recount_score = {}



for data in datas:
    role_name = data['role_name']

    for short_name in long_name.keys():
        n = -1
        res = []
        table_name = f'{short_name}_table'
        for k in data[table_name].keys():
            v = data[table_name][k]
            if n == -1:
                n = len(v)
                res = [-999 for _ in range(n)]
            for i, a in enumerate(v):
                if a!='null' and res[i]==-999:
                    res[i] = a

            for i in range(len(res)):
                if isinstance(res[i], str):
                    res[i] = int(res[i])

        # n = len()
        average = float(sum(res))/ float(len(res))
        recount_score[(role_name, short_name)] = average
        print(res)
        break
    break

[-5, 5, 2, -2, 2]


In [9]:
print(role_name)
print(long_name[short_name])
print(res)

汤师爷
neuroticism
[-5, 5, 2, -2, 2]


In [ ]:
print('角色', end = ' | ')

for key in long_name.keys():
    print(long_name[key], end = ' | ')

print()

print('|-|',end = '')

for key in long_name.keys():
    print('-', end = ' | ')

print()


for data in datas:
    role_name = data['role_name']

    print('|', role_name, end = ' | ')

    for key in long_name.keys():
        # major_score = data[key]['score']
        major_score = recount_score[(role_name, key)]

        print('%.3f' % major_score, end = ' | ')

    print()



角色 | neuroticism | extraversion | openness | agreeableness | conscientiousness | 
|-|- | - | - | - | - | 
| 汤师爷 | 0.400 | -2.000 | -0.286 | -0.250 | 0.000 | 
| 慕容复 | -3.800 | 1.167 | 1.286 | -0.750 | 3.375 | 
| 李云龙 | -2.000 | -0.667 | -2.857 | -1.000 | 1.500 | 
| Luna | -0.400 | 2.000 | 5.000 | 2.250 | 0.125 | 
| 王多鱼 | -2.600 | 3.500 | 0.714 | -1.500 | -1.000 | 
| Ron | -0.600 | -0.833 | -1.429 | 1.125 | -0.750 | 
| 鸠摩智 | -2.000 | -2.000 | 1.333 | -0.250 | 2.500 | 
| Snape | -1.200 | -2.500 | -2.000 | -2.000 | 1.750 | 
| 凉宫春日 | -2.600 | 2.833 | 3.714 | -1.500 | 1.500 | 
| Malfoy | -2.000 | -2.000 | -2.000 | -2.000 | 1.750 | 
| 虚竹 | 1.200 | -2.500 | 0.714 | 4.250 | 0.875 | 
| 萧峰 | -2.000 | -0.167 | -0.857 | 1.500 | 3.375 | 
| 段誉 | -1.800 | 2.167 | 1.714 | 4.625 | 0.875 | 
| Hermione | -2.000 | 0.000 | 4.571 | 2.000 | 4.250 | 
| Dumbledore | -2.000 | 2.500 | 4.571 | 3.125 | 2.625 | 
| 王语嫣 | 1.000 | -0.333 | 2.857 | 3.000 | 2.750 | 
| Harry | -0.400 | -0.500 | -0.286 | 2.500 | 3.750 | 
| 

In [ ]:
fname_eval = '/content/drive/MyDrive/Psy/psy_eval_gpt.txt'

In [ ]:
import json

datas = []

with open(fname_eval, encoding='utf-8') as f:
    for line in f:
        data = json.loads(line)
        datas.append(data)

In [ ]:
save_name = '/content/drive/MyDrive/Psy/psy_eval_gpt_with_id.txt'

In [ ]:
eval_id = 0
with open(save_name, 'w', encoding='utf-8') as f:
    for data in datas:
        data['eval_batch_id'] = eval_id
        json.dump(data, f, ensure_ascii=False)
        f.write('\n')
        eval_id += 1



fname_eval = '/content/drive/MyDrive/Psy/psy_eval_gpt.txt'

import json

datas = []

with open(fname_eval, encoding='utf-8') as f:
    for line in f:
        data = json.loads(line)
        datas.append(data)

save_name = '/content/drive/MyDrive/Psy/psy_eval_gpt_with_id.txt'

import json


with open(save_name, 'w', encoding='utf-8') as f:


我现在想在每条data中增加一个字段 eval_batch_id 从0开始计数

保存在save_name, ensure_assic = False

请用python为我实现